tests/net/dhcpv4/server/src/main.c

```diff
@@ -0,0 +1,867 @@
+#include <zephyr/net/net_ip.h>
+#include <zephyr/net/dhcpv4.h>
+#include <zephyr/net/dhcpv4_server.h>
+
+#include "dhcpv4.h"
+#include "ipv4.h"
+#include "udp_internal.h"
+
+/* 00-00-5E-00-53-xx Documentation RFC 7042 */
+static uint8_t server_mac_addr[] = { 0x00, 0x00, 0x5E, 0x00, 0x53, 0x01 };
+static uint8_t client_mac_addr[] = { 0x00, 0x00, 0x5E, 0x00, 0x53, 0x02 };
+
+static struct in_addr server_addr = { { { 192, 0, 2, 1 } } };
+static struct in_addr netmask = { { { 255, 255, 255, 0 } } };
+static struct in_addr test_base_addr = { { { 192, 0, 2, 10 } } };
+
+/* Only to test Inform. */
+static struct in_addr client_addr_static = { { { 192, 0, 2, 2 } } };
+
+typedef void (*test_dhcpv4_server_fn_t)(struct net_if *iface,
+					struct net_pkt *pkt);
+
+
+static struct test_dhcpv4_server_ctx {
+	struct net_if *iface;
+	struct k_sem test_proceed;
+	struct net_pkt *pkt;
+	struct in_addr assigned_ip;
+
+	/* Request params */
+	const char *client_id;
+	int lease_time;
+	bool broadcast;
+} test_ctx;
+
+struct test_lease_count {
+	int reserved;
+	int allocated;
+	int declined;
+};
+
+#define CLIENT_ID_1 "client1"
+#define CLIENT_ID_2 "client2"
+#define NO_LEASE_TIME -1
+#define TEST_XID 0x12345678
+
+#define TEST_TIMEOUT K_MSEC(100)
+
+static void server_iface_init(struct net_if *iface)
+{
+	net_if_set_link_addr(iface, server_mac_addr, sizeof(server_mac_addr),
+			     NET_LINK_ETHERNET);
+
+	test_ctx.iface = iface;
+
+	(void)net_if_ipv4_addr_add(iface, &server_addr, NET_ADDR_MANUAL, 0);
+	(void)net_if_ipv4_set_netmask(iface, &netmask);
+}
+
+static int server_send(const struct device *dev, struct net_pkt *pkt)
+{
+	test_ctx.pkt = pkt;
+	net_pkt_ref(pkt);
+
+	k_sem_give(&test_ctx.test_proceed);
+
+	return 0;
+}
+
+static struct dummy_api server_if_api = {
+	.iface_api.init = server_iface_init,
+	.send = server_send,
+};
+
+NET_DEVICE_INIT(server_iface, "server_iface", NULL, NULL, NULL, NULL,
+		CONFIG_KERNEL_INIT_PRIORITY_DEFAULT, &server_if_api,
+		DUMMY_L2, NET_L2_GET_CTX_TYPE(DUMMY_L2), NET_IPV4_MTU);
+
+static const uint8_t cookie[4] = { 0x63, 0x82, 0x53, 0x63 };
+
+static void test_pkt_free(void)
+{
+	if (test_ctx.pkt != NULL) {
+		net_pkt_unref(test_ctx.pkt);
+		test_ctx.pkt = NULL;
+	}
+}
+
+static void client_prepare_test_msg(
+	const struct in_addr *src_addr, const struct in_addr *dst_addr,
+	enum net_dhcpv4_msg_type type, const struct in_addr *server_id,
+	const struct in_addr *requested_ip, const struct in_addr *ciaddr)
+{
+	struct dhcp_msg msg = { 0 };
+	uint8_t empty_buf[SIZE_OF_FILE] = { 0 };
+	struct net_pkt *pkt;
+
+	pkt = net_pkt_alloc_with_buffer(test_ctx.iface, NET_IPV4_MTU, AF_INET,
+					IPPROTO_UDP, K_FOREVER);
+	zassert_not_null(pkt, "Failed to allocate packet");
+
+	net_pkt_set_ipv4_ttl(pkt, 1);
+
+	if (net_ipv4_create(pkt, src_addr, dst_addr) ||
+	    net_udp_create(pkt, htons(DHCPV4_CLIENT_PORT),
+			   htons(DHCPV4_SERVER_PORT))) {
+		zassert_true(false, "Failed to create IPv4/UDP headers");
+	}
```

Comment:
Even shorter and more descriptive, but I also agree we don't have to re-assert stuff that is never expected to fail.

```c
	zassert_ok(net_ipv4_create(pkt, src_addr, dst_addr),
		  "Failed to create IPv4 header");
	zassert_ok(net_udp_create(pkt, htons(DHCPV4_CLIENT_PORT)),
		  "Failed to create UDP header");
```

Reply:
Agreed, updated as suggested.